### PR TITLE
[TS-1225] Session managament based on NextExpectedSeqNum

### PR DESCRIPTION
**Pull request by isengrims**

### CODE REVIEW

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 1]
```diff
@@ -628,24 +740,19 @@ public void onClose(@NotNull IChannel channel) {
...
-
-        if (enabled.get()) {
-            channel.send(Unpooled.wrappedBuffer(logout.toString().getBytes(StandardCharsets.UTF_8)), Collections.emptyMap(), null, IChannel.SendMode.MANGLE);
-        }
-        disconnectRequest = executorService.schedule(() -> enabled.set(false), settings.getDisconnectRequestDelay(), TimeUnit.SECONDS);
+        sendLogout();
     }
 
-    private void setHeader(StringBuilder stringBuilder, String msgType, Integer seqNum) {
+    private void setHeader(StringBuilder stringBuilder, String msgType, Integer seqNum, Boolean possDup) {
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:746]: > Boolean possDup

It's always `false`

File: src/main/java/com/exactpro/th2/FixHandlerSettings.java [hunk 2]
```diff
@@ -46,6 +49,9 @@ public class FixHandlerSettings implements IHandlerSettings {
     private String passwordEncryptAlgorithm = "RSA";
     private Boolean resetSeqNumFlag = false;
     private Boolean resetOnLogon = false;
+    private Boolean maintainSessionBasedOnNextExpectedSeqNumber = false;
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandlerSettings.java:52]: Maybe rename to `useNextExpectedSeqNum`/`checkNextExpectedSeqNum`?

File: src/main/java/com/exactpro/th2/FixHandlerSettings.java [hunk 3]
```diff
@@ -46,6 +49,9 @@ public class FixHandlerSettings implements IHandlerSettings {
...
+    private Boolean isSaveAdminMessages = false;
+    private String startOfADayTime;
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandlerSettings.java:54]: > private String startOfADayTime

Maybe rename to `sessionStartTime`?
And we should also have `sessionEndTime` setting too
Also I guess we can change its type to `LocalTime` - Jackson should be able to handle that (maybe certain annotations will be required)

File: src/main/java/com/exactpro/th2/SequenceHolder.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2022-2023 Exactpro (Exactpro Systems Limited)
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/SequenceHolder.java:2]:
```suggestion
 * Copyright 2023 Exactpro (Exactpro Systems Limited)
```

File: src/main/java/com/exactpro/th2/util/Util.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2022-2023 Exactpro (Exactpro Systems Limited)
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/util/Util.java:2]:
```suggestion
 * Copyright 2023 Exactpro (Exactpro Systems Limited)
```

File: src/main/java/com/exactpro/th2/util/Util.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            int firstLine = Integer.parseInt(br.readLine());
+            int secondLine = Integer.parseInt(br.readLine());
+            return new SequenceHolder(firstLine, secondLine);
+        } catch (IOException e) {
+            throw new IllegalStateException("Error while reading sequence file " + filePath, e);
+        }
+    }
+
+    public static void writeSequences(int msgSeqNum, int serverSeqNum, String filePath) throws IOException {
+        BufferedWriter bw = new BufferedWriter(new FileWriter(filePath));
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/util/Util.java:38]: > BufferedWriter bw = new BufferedWriter(new FileWriter(filePath))

Should be wrapped in "try-with-resources"

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 7]
```diff
@@ -606,6 +698,26 @@ public void sendLogon() {
...
+        }
+        disconnectRequest = executorService.schedule(() -> {
+            enabled.set(false);
+            try {
+                Util.writeSequences(msgSeqNum.get(), serverMsgSeqNum.incrementAndGet(), settings.getStateFilePath());
+            } catch (IOException e) {
+                if (LOGGER.isErrorEnabled()) {
+                    LOGGER.error("Sequence number were not saved: clientSeq - {}, serverSeq - {}", msgSeqNum.get(), serverMsgSeqNum.get());
+                }
+            }
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:717]: Why are we delaying it? This delay could prevent us from saving sequences in time when this method is called from `close` method

**Review by ivandruzhinin — CHANGES_REQUESTED** [review]

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 8]
```diff
@@ -168,10 +161,8 @@ public FixHandler(IHandlerContext context) {
...
             serverMsgSeqNum = new AtomicInteger(sequences.getServerSeq());
         }
-        if(settings.getStartOfADayTime() != null) {
-            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm:ss").withZone(ZoneOffset.UTC);
-            Instant instant = Instant.from(formatter.parse(settings.getStartOfADayTime()));
-            resetTime = LocalDateTime.ofInstant(instant, ZoneId.of("UTC")).toLocalTime();
+        if(settings.getSessionStartTime() != null) {
+            LocalTime resetTime = settings.getSessionStartTime();
             LocalDateTime now = LocalDateTime.now();
             LocalDateTime scheduleTime;
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:167]: I guess we still should treat startTime/endTime as UTC

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 9]
```diff
@@ -180,9 +171,20 @@ public FixHandler(IHandlerContext context) {
                 scheduleTime = now.plusDays(1).with(resetTime);
             }
             long time = scheduleTime.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli() - System.currentTimeMillis();
-            executorService.schedule(this::reset, time, TimeUnit.MILLISECONDS);
-        } else {
-            resetTime = LocalTime.now();
+            executorService.scheduleAtFixedRate(() -> reset(settings.getSessionEndTime() == null), time, 24, TimeUnit.HOURS);
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:174]: > settings.getSessionEndTime() == null

I think we should make `sessionEndTime` mandatory if `sessionStartTime` is present

> reset(settings.getSessionEndTime() == null)

We shouldn't not only reset seqnums at start time but also send login

> time, 24, TimeUnit.HOURS

`time` is in milliseconds

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 10]
```diff
@@ -439,34 +439,38 @@ private void handleResendRequest(ByteBuf message) {
...
+            ByteBuf storedMsg = outgoingMessages.get(i);
+            if (storedMsg == null) {
+                StringBuilder heartbeat = new StringBuilder();
+                setHeader(heartbeat, MSG_TYPE_HEARTBEAT, i);
+                setChecksumAndBodyLength(heartbeat);
+                channel.send(Unpooled.wrappedBuffer(heartbeat.toString().getBytes(StandardCharsets.UTF_8)), Collections.emptyMap(), null, SendMode.MANGLE);
+            } else {
+                if (LOGGER.isInfoEnabled()) LOGGER.info("Resending message: {}", storedMsg.toString(US_ASCII));
+                FixField sendingTime = findField(storedMsg, SENDING_TIME_TAG);
+                sendingTime.insertNext(POSS_DUP_TAG, IS_POSS_DUP);
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:464]: Why do we always set `PossDupTag` to `Y`?

File: src/main/java/com/exactpro/th2/FixHandlerSettings.java [hunk 11]
```diff
@@ -49,9 +52,17 @@ public class FixHandlerSettings implements IHandlerSettings {
     private String passwordEncryptAlgorithm = "RSA";
     private Boolean resetSeqNumFlag = false;
     private Boolean resetOnLogon = false;
-    private Boolean maintainSessionBasedOnNextExpectedSeqNumber = false;
+    private Boolean useNextExpectedSeqNum = false;
     private Boolean isSaveAdminMessages = false;
-    private String startOfADayTime;
+
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "HH:mm:ss.SSS")
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandlerSettings.java:58]: > HH:mm:ss.SSS

Second and millisecond parts can be optional

**Review by ivandruzhinin — CHANGES_REQUESTED** [review]

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 12]
```diff
@@ -314,7 +314,7 @@ public Map<String, String> onIncoming(@NotNull IChannel channel, @NotNull ByteBu
                 if (LOGGER.isInfoEnabled()) LOGGER.info("Logon received - {}", message.toString(US_ASCII));
                 boolean connectionSuccessful = checkLogon(message);
                 if (connectionSuccessful) {
-                    msgSeqNum.incrementAndGet();
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java]: Why it was removed?

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 13]
```diff
@@ -676,7 +676,7 @@ public void sendLogon() {
         else reset = settings.getResetOnLogon();
         if (reset) msgSeqNum.getAndSet(0);
 
-        setHeader(logon, MSG_TYPE_LOGON, msgSeqNum.get() + 1);
+        setHeader(logon, MSG_TYPE_LOGON, msgSeqNum.incrementAndGet());
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:679]: > msgSeqNum.incrementAndGet()

MsgSeqNum should be incremented only after successful login

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 14]
```diff
@@ -606,6 +703,28 @@ public void sendLogon() {
...
+            try {
+                Util.writeSequences(msgSeqNum.get(), serverMsgSeqNum.incrementAndGet(), settings.getStateFilePath());
+            } catch (IOException e) {
+                if (LOGGER.isErrorEnabled()) {
+                    LOGGER.error("Sequence number were not saved: clientSeq - {}, serverSeq - {}", msgSeqNum.get(), serverMsgSeqNum.get());
+                }
+            }
+        }
+        disconnectRequest = executorService.schedule(() -> {
+            enabled.set(false);
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:724]: Why do we need to delay it? Can we set it right away?

**Review by ivandruzhinin — CHANGES_REQUESTED** [review]

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 15]
```diff
@@ -169,27 +168,25 @@ public FixHandler(IHandlerContext context) {
...
-            if(now.with(resetTime).isAfter(now)) {
+            ZonedDateTime scheduleTime = now.with(resetTime);
+            if(scheduleTime.isBefore(now)) {
                 scheduleTime = now.with(resetTime);
-            } else {
-                scheduleTime = now.plusDays(1).with(resetTime);
             }
-            long time = now.until(scheduleTime, ChronoUnit.HOURS);
-            executorService.scheduleAtFixedRate(this::reset, time, 24, TimeUnit.HOURS);
+            long time = now.until(scheduleTime, ChronoUnit.MINUTES);
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:175]: > ChronoUnit.MINUTES

Let's narrow it down to seconds

File: src/main/java/com/exactpro/th2/util/FileDeserializer.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+package com.exactpro.th2.util;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+
+import java.io.File;
+import java.io.IOException;
+
+public class FileDeserializer extends StdDeserializer<File> {
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/util/FileDeserializer.java:25]: Jackson has built-in deserialzier for `File` (see `FromStringDeserializer`) which will be applied automatically

File: README.md [hunk 17]
```diff
@@ -41,6 +41,11 @@ This microservice allows sending and receiving messages via FIX protocol
 + *disconnectRequestDelay* - the interval for the shutdown request
 + *resetSeqNumFlag* - resetting sequence number in initial Logon message (when conn started)
 + *resetOnLogon* - resetting the sequence number in Logon in other cases (e.g. disconnect)
++ *stateFilePath* - path to file where sequences will be saved to use with next login attempts. It is useful when acceptor is not support sequence reset. (`nullable`)
```
Comment (ivandruzhinin) [inline README.md:44]:
```suggestion
+ *stateFilePath* - path to file where sequences will be saved to use with next login attempts. It is useful when acceptor does not support sequence reset. (`nullable`)
```

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 18]
```diff
@@ -164,16 +164,22 @@ public FixHandler(IHandlerContext context) {
             msgSeqNum = new AtomicInteger(sequences.getClientSeq());
             serverMsgSeqNum = new AtomicInteger(sequences.getServerSeq());
         }
+
+        LOGGER.info("Initial sequences are: client - {}, server - {}", msgSeqNum.get(), serverMsgSeqNum.get());
+
         if(settings.getSessionStartTime() != null) {
             Objects.requireNonNull(settings.getSessionEndTime(), "Session end is required when session start is presented");
+            if(settings.getSessionStartTime().isBefore(settings.getSessionEndTime())) {
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:172]: > if(settings.getSessionStartTime().isBefore(settings.getSessionEndTime())) {

This'll have false positives for some cases (e.g. when start is at `08:00` and end is at `02:00` (of the next day))

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 19]
```diff
@@ -395,6 +402,8 @@ private void resetSequence(ByteBuf message) {
 
         if (seqNumValue != null && (gapFillFlagValue == null || requireNonNull(gapFillFlagValue.getValue()).equals("N"))) {
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:403]: > if (seqNumValue != null && (gapFillFlagValue == null || requireNonNull(gapFillFlagValue.getValue()).equals("N")))

This is fully covered by the next if-block

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 20]
```diff
@@ -164,16 +164,22 @@ public FixHandler(IHandlerContext context) {
...
             ZonedDateTime now = ZonedDateTime.now(ZoneOffset.UTC);
             ZonedDateTime scheduleTime = now.with(resetTime);
             if(scheduleTime.isBefore(now)) {
-                scheduleTime = now.with(resetTime);
+                scheduleTime = now.plusDays(1).with(resetTime);
             }
-            long time = now.until(scheduleTime, ChronoUnit.MINUTES);
-            executorService.scheduleAtFixedRate(this::reset, time, 24 * 60, TimeUnit.MINUTES);
+            long time = now.until(scheduleTime, ChronoUnit.SECONDS);
+            executorService.scheduleAtFixedRate(this::reset, time, 24 * 60 * 60, TimeUnit.MINUTES);
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:182]: > 24 * 60 * 60

Can be extracted into a constant

File: src/main/java/com/exactpro/th2/FixHandler.java [hunk 21]
```diff
@@ -133,14 +149,52 @@ public class FixHandler implements AutoCloseable, IHandler {
...
+            LocalTime resetTime = settings.getSessionEndTime();
+            ZonedDateTime now = ZonedDateTime.now(ZoneOffset.UTC);
+            ZonedDateTime scheduleTime = now.with(resetTime);
+
+            if(scheduleTime.isBefore(now)) {
+                scheduleTime = now.plusDays(1).with(resetTime);
+            }
+
+            long time = now.until(scheduleTime, ChronoUnit.SECONDS);
+            executorService.scheduleAtFixedRate(this::close, time, 24 * 60 * 60, TimeUnit.MINUTES);
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/FixHandler.java:195]: We should also block message-send attempts when session is inactive

**Review by ivandruzhinin — APPROVED** [review]

**Comment by isengrims** [comment]
Already merged into dev-version-0 and dev-version-1

**Pull request closed without merging**